sql-plugin/src/main/scala/org/apache/spark/sql/rapids/complexTypeExtractors.scala

```diff
@@ -210,4 +202,49 @@ case class GpuArrayContains(left: Expression, right: Expression)
     lhs.getBase.listContainsColumn(rhs.getBase)
 
   override def prettyName: String = "array_contains"
-}
\ No newline at end of file
+}
+
+/** Core static methods for GetArrayItem and ElementAt
+ */
+object GetArrayItemUtil extends Arm{
+  def evalColumnar(array: GpuColumnVector, ordinal: Scalar, dataType: DataType,
+                   zeroIndexed: Boolean): ColumnVector = {
+    // for array index use case, index starts at 0
+    if (zeroIndexed) {
+      // Need to handle negative indexes...
+      if (ordinal.isValid && ordinal.getInt >= 0) {
+        array.getBase.extractListElement(ordinal.getInt)
+      } else {
+        withResource(Scalar.fromNull(
+          GpuColumnVector.getNonNestedRapidsType(dataType))) { nullScalar =>
+          ColumnVector.fromScalar(nullScalar, array.getRowCount.toInt)
+        }
+      }
+    } else {
+      // for element_at use case, index starts at 1
+      if (ordinal.isValid) {
+        if (ordinal.getInt > 0) {
+          // SQL 1-based index
+          array.getBase.extractListElement(ordinal.getInt - 1)
+        } else if (ordinal.getInt == 0) {
+          throw new ArrayIndexOutOfBoundsException("SQL array indices start at 1")
+        } else {
+          array.getBase.extractListElement(ordinal.getInt)
+        }
+      } else {
+        withResource(Scalar.fromNull(
```

Comment:
can we future proof this a little? We already have the data type this is supposed to be, it is just the Spark version of it.

```
withResource(GpuScalar.from(null, dataType))  { nullScalar =>
```

This also goes for above, but if we are going to revert things so they are separate again you only need to do this for new code.

Reply:
Sure, will do it when the GpuScalar PR gets merged.

Reply:
Updated.